### ISSUE TITLE
New Scan API

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pytensor#1940]
### Description

**⚠️ The implementation of this issue is reserved for core developers, feedback welcome from anyone⚠️**

This was lightly discussed internally. We suggest adding a new user-facing Scan API similar to JAX's, with signature:

```python
def step_fn(carry, x):
  return new_carry, None
  # with traced outputs (old outputs_info=[None])
  return new_carry, y, 
  # optional break_cond, that turns it into a while scan
  return new_carry, y, break_cond
  
final, ys, done = scan(step_fn, init, xs, n_steps)
# done only if there was a break_cond, it denotes whether the break_cond ever become true
# (as we still require a maximum number of n_steps in the inner scan, or n_steps could be zero)
```

* init/carry, sequences, and y can be a simple pytree (nested tuple/dictionary), to avoid flat signature mess users need to navigate these days.
* carry/new_carry pytree must have the same structure and root types
* Internally we keep using the existing ScanOp, which requires flattenning everything, and ordering things in a specific way, but the user only interacts directly with the structured inputs and outputs
* The existing user-facing scan function (that helps build a ScanOp) would still exist with a deprecation warning. New one will need a new name or new module location. maybe call it `pytensor.loop`?

## Taps
Taps is one of the largest sources of complexity of Scan, but arguably quite aligned with memory layout we want in the end (i.e., a single outer buffer whose last n entries can be accessed in the inner function)

The user could naively implement taps like:
```python
init = (atm2, atm1)

def step_fn(carry, x):
  atm2, atm1 = carry
  at = f(atm2, atm1, x)
  next_carry = (atm1, at)
```
But internally we want to convert xtm2, xtm1 to a tap outputs_info in the regular scan. Graph analysis is perhaps too tricky for this, so I suggest implementing an helper.

```python
init = scan.taps(a, taps=[-2, -1])  # if taps=None, default to all the entries of x leading dimension, if length is statically known

def step_fn(carry, x):
  atm2, atm1 = carry
  at = f(atm2, atm1, x)
  next_carry = carry.push_tap(xt)
```

* scan.taps creates a special class that new pytensor scan recognizes. They must be at the bottom of the pytree, and can be composed with it (that is stuff like `carry=(taps(a), b)` is also fine).
  * We can raise informative errors if `carry.push_tap` wasn't called or was called multiple times. 
* Scan sequences also allow taps (with positive integers). Can also composed at the bottom of the `xs` pytree.
  * We forbid `carry.push_tap` in these
* If we find graph analysis is sufficiently robust, we can always deprecate later down the road

## User benefits
* Simpler syntax
  * It's currently very hard to get the internal function signature correctly when you have many sequences and outputs with taps. Codegen scans with arbitrary functions are particularly hard, and already require some sort of flatenning/unflatening ad-hoc logic to keeps things manageable
* Easier adoption
  * Specially if you already know JAX. I feel devs are much more at ease writing jax scans than they are at writing pytensor scans, even though they are conceptually quite similar
  * Having a python helper / class with proper docstrings / introspectability for `scan.taps` is a huge plus. I always have to go and check how the dictionary that's required know looks like (what are the keys).
* Consistency for updates of variables that can't be traced, like RNGs. The deprecated `updates` signature was mostly for this. The new proposed API with `return_updates=False`, puts these outputs together with regular recurring outputs with the gotcha that you only get back the last state, which is inconsistent. For regular outputs you still need to do [-1] if you only want the last one.
  * For types that can't be traced, Scan will be happy to handle them in the `init/carry/final` side of the Scan, and will raise an informative error if you try to trace it as a `y` 
* Valid state for n_steps=0. Because users are only given a "trace" in the current implementation, there's no clean way to get last_state if n_steps=0. With the init/final design, this is valid. last_state = init_state. No risky `trace[init:][-1]` business. (There's some resolved discussion in https://github.com/pymc-devs/pymc/pull/7380#discussion_r2510579256)
  * It's still invalid for traced outputs (aka the y in the example above), but you usually want the whole sequence for those, and an empty sequence is well defined 
    * (except for the little gotcha of not always knowing the exact zero shape it should have, `shape=(0, 5)` is different than `shape=(0, 2, 3)`)

## Internal benefits
* Most cases of `scan_save_mem` become trivial to analyze. We mostly need to distinguish between `traced` scan and `final_state` scan. The cases where a user wants the last n states, with n > 1, are exceedingly rare. The cases where the user wants the first n < N steps are also rare, sometimes in autodiff, when the gradient doesn't depend on the output quantity, the very last step can be cut. I reckon it's fine to lose this fine-grained optimization while focusing on the main one (whole scan or only last state).
  * `scan_save_mem` is one of the most complicated rewrites we have, and it's mostly due to bad initial semantics for Scan
  *  This also reduces the need for `local_subtensor_merge` somewhat which is a source of graph explosion and rewrite slowdown: https://github.com/pymc-devs/pytensor/issues/112

**Comment by jessegrabowski** [comment]
First of all, I am broadly in favor of an API refactor. I think something like this is a lot more ergonomic. Some questions:

1) Do we care about the appearance of being "baby jax" by copying their scan API? We already compile to jax, and we start making our code look like jax, at some point people ask "why not just use jax". I know there's basically nobody else with a scan Op to look to for guidance, but we might want to at least explore some alternatives
2) You mention pytrees, but we don't actually have support for those anywhere else. Would that require a new effort more broadly in the codebase? Also pytrees are another jax concept, so back to (1)
3) Does your taps proposal support non-consecutive taps? The example I always give is seasonal ARIMA so init = taps(x, [-2, -1]). 
3a) If we're doing this I think it's a good opportunity to break with the word "Taps". I don't understand what it means anyway. It would be nicer if we had `lag` and `lead` operators, each with a `shift=[1, 2]`. So `lag(x, [1, 2])` -> `taps(x, [-1, -2])`, and `lead(x, [1, 2])` -> `taps(x, [1, 2])`. This aligns with timeseries literature and it's more clear what everything is doing. I would even allow positive and negative shift argument in the `lag` argument (but they both do the same thing).
3b) If you insist on keeping taps, the signature `taps(x, taps)` is ugly, why is it both the function name and the shift argument?
3c) I dislike the boilerplate associated with push_taps. This reminds me of collect_default_updates, which I was very thankful we did away with. We should really really fight to do this step via graph inspection if at all possible
4) It's funny that you say the updates API is mostly for RNGs. It was originally intended for RNN trainable parameters I assume :)
5) re: loop vs scan, I don't love the name loop. We could consider calling the existing scan something like `internal_scan` or `canonical_scan` and keep the public-facing function called `scan`? Users can `from scan import canonical_scan as scan` for backwards compat.

**Comment by ricardoV94** [comment]
> Do we care about the appearance of being "baby jax" by copying their scan API? 

No. Why should we limit ourselves away from good ideas?

>  We already compile to jax, and we start making our code look like jax, at some point people ask "why not just use jax". 

That has already happened. So far the reason is we want different backends, and to provide easy graph manipulation. Also jax is too limiting with static shapes, doesn't do great on CPU, not very extensible. If JAX gets better than us on all those fronts sure, why shouldn't people use it? 

Again, having worse syntax is not how we win. And pytensor scan is something nobody groks or loves (prove me wrong).

> I know there's basically nobody else with a scan Op to look to for guidance, but we might want to at least explore some alternatives

https://onnx.ai/onnx/operators/onnx__Scan.html

> You mention pytrees, but we don't actually have support for those anywhere else. 

I want us to move to supporting those as input/outputs of a pytensor function. It's really cumbersome to have to work only with flat input/outputs. For instance, variable shapes should output tuples not 1d vector, but then you can't `function([x], x.shape)`

However, doing the flattening/unflattening in the function would cause evaluation overhead (unless we move that to the compiled code itself), so I'm not proposing it yet. 

This has no runtime overhead, it's just syntactic sugar to build the Scan Op (and the rest of the graph afterwards)

> Does your taps proposal support non-consecutive taps? The example I always give is seasonal ARIMA so init = taps(x, [-2, -1]).

Yes. That example is consecutive though?

> If we're doing this I think it's a good opportunity to break with the word "Taps". I don't understand what it means anyway. It would be nicer if we had lag and lead operators, each with a shift=[1, 2]. So lag(x, [1, 2]) -> taps(x, [-1, -2]), and lead(x, [1, 2]) -> taps(x, [1, 2]). This aligns with timeseries literature and it's more clear what everything is doing. I would even allow positive and negative shift argument in the lag argument (but they both do the same thing).

I don't like the conversion from positive numbers to negative indices. I don't like two functions because of the sign. 

I also don't like taps (which comes from signal processing IIRC). What about `shift(var, by=[-2, -1]`?

> I dislike the boilerplate associated with push_taps. This reminds me of collect_default_updates, which I was very thankful we did away with. We should really really fight to do this step via graph inspection if at all possible

I don't like the mit-sot syntax, where you get n numbers and only return 1. I find it hard to map it to the input when reading a scan function. So I thought being explicit about. 

We can go more simple with `push`? 

I also want the mental model to be almost what a naive hand-rolled taps (shifts?) would be: `update = (*input_taps[1:], new_value)`.

> It's funny that you say the updates API is mostly for RNGs. It was originally intended for RNN trainable parameters I assume :)

Internally, updates/shared variables only really exist(ed) for RNGs. regular tensors are simply converted to sit-sot internally. I've renamed that to `untraceable_sit_sot` recently.


> re: loop vs scan, I don't love the name loop. We could consider calling the existing scan something like internal_scan or canonical_scan and keep the public-facing function called scan? Users can from scan import canonical_scan as scan for backwards compat.

You love scan? I would bet any beginnre user reaching for scan was looking for "how to loop" in PyTensor. That's how we name it in the docs as well 

<img width="702" height="102" alt="Image" src="https://github.com/user-attachments/assets/53e63fe7-e5f3-4622-a8f6-b22caffd9b50" />

Scan is too cute (and I've done some haskell in the past, it's not fear from the concept)

Scan is also distinguished from reduce/fold because it returns all intermediate results (the traces), which this syntax does not (unless you add them as ys). So it's a less apt name. Sure, we still trace everything when we autodiff, but that's an implementation detail, not the user-facing behavior.

**Comment by JiwaniZakir** [comment]
The proposed API closely mirrors `jax.lax.scan`, which should lower the barrier to entry for users familiar with JAX and reduce the flat-signature complexity that currently makes pytensor's scan difficult to use. The key implementation challenge will be the pytree flattening/unflattening layer that bridges the new structured API to the existing `ScanOp`, which expects inputs ordered as `[sequences, outputs_info, non_sequences]` — this translation logic will likely live somewhere adjacent to `pytensor/scan/utils.py` or as a new wrapper in `pytensor/scan/__init__.py`. The taps helper is a particularly important UX detail: without it, users who naively encode taps as tuple carries will produce a suboptimal graph (redundant storage of intermediate steps), so the helper needs to detect the tap pattern and rewrite it into proper `outputs_info` entries with the correct negative-integer tap indices before the `ScanOp` is constructed. The `break_cond` returning a `done` flag also implies the new wrapper needs to handle the `until` condition that maps to `pytensor.scan.op.until`, and the semantics of `done` (whether the condition was *ever* true vs. whether it terminated early) should be clarified to avoid confusion with the current `until` behavior.

**Comment by ricardoV94** [comment]
@JiwaniZakir we all have LLMs, your comment adds nothing. Please refrain from that kind of interaction.
